dashboard/test/models/concerns/user/level_progressable_test.rb

```diff
@@ -0,0 +1,671 @@
+        # User's most recent progress is on last level in script. There's nothing
+        # following it, so just return to the last level
+        _(subject.chapter).must_equal script_level.chapter
+      end
+    end
+
+    context 'when most recent level not a progression level' do
+      let(:user) {create :user}
+      let(:script) {create :script}
+      let(:lesson_group) {create :lesson_group, script: script}
+      let(:lesson) {create :lesson, script: script, lesson_group: lesson_group}
+      let!(:script_levels) do
+        [
+          create(:script_level, script: script, lesson: lesson, levels: [create(:maze)]),
+          create(:script_level, script: script, lesson: lesson, levels: [create(:unplugged)]),
+          create(:script_level, script: script, lesson: lesson, levels: [create(:unplugged)]),
+          create(:script_level, script: script, lesson: lesson, levels: [create(:maze)])
+        ]
+      end
+      before do
+        create :user_script, user: user, script: script
+
+        UserLevel.create(
+          user: user,
+          level: script_levels[1].level,
+          script: script,
+          attempts: 1,
+          best_result: Activity::MINIMUM_PASS_RESULT
+        )
+      end
+
+      it 'returns the the next uncompleted progression level' do
+        # User's most recent progress is on unplugged level, that is followed by another
+        # unplugged level. We should end up at the first non unplugged level
+        _(subject.chapter).must_equal 4
+      end
+    end
+
+    context 'when last updated user_level is inside a level group' do
+      let(:user) {create :user}
+      let(:script) {create :script}
+      let(:sub_level_name) {'sublevel1'}
+      let(:lesson_group) {create :lesson_group, script: script}
+      let(:lesson) {create :lesson, script: script, lesson_group: lesson_group}
+
+      let!(:sub_level1) {create :text_match, name: sub_level_name}
+      let!(:level_group) {create_level_group(sub_level_name)}
+
+      before do
+        create :script_level, script: script, levels: [level_group], lesson: lesson
+        create :user_script, user: user, script: script
+
+        # Create a UserLevel for our level_group and sublevel, the sublevel is more recent
+        UserLevel.create(
+          user: user,
+          level: level_group,
+          script: script,
+          attempts: 1,
+          best_result: Activity::MINIMUM_PASS_RESULT,
+          updated_at: Time.now - 1
+        )
+
+        UserLevel.create(
+          user: user,
+          level: sub_level1,
+          script: script,
+          attempts: 1,
+          best_result: Activity::MINIMUM_PASS_RESULT,
+          updated_at: Time.now
+        )
+      end
+      it 'returns a non-nil script level' do
+        _(subject).wont_be_nil
+      end
+    end
+  end
+
+  describe '#completed_progression_levels?' do
+    subject(:completed_progression_levels?) {user.completed_progression_levels?(script)}
```

Comment:
This is mostly for my own RSpec education: is there a particular reason to use the `subject(:name) {}` syntax here and `_name` below, rather than using `subject {}` here and `_(subject)` below?

Reply:
We prefer to use `subject(:name) {} ` and `_name` for clarity and readability.
We side with explicitly suggesting what’s being tested in each assertion, rather than a generic `_(subject)`.

Reply:
I do like explicit over implicit as a general principle! In that case, my question becomes: is there a reason we're using a generic `_(subject)` for all the _other_ tests? :smile:

Reply:
Ahh! gotcha, yeah, I am changing those.
I used the generic `_(subject)` as a first draft of the tests since the tests access properties of the level in the assertions.
With all the tests that needed to be rewritten, I forgot to set the explicit subject afterwards.